source/common/upstream/load_balancer_impl.cc

```diff
@@ -513,13 +559,28 @@ HostConstSharedPtr LeastRequestLoadBalancer::unweightedHostPick(const HostVector
   }
 }
 
-HostConstSharedPtr RandomLoadBalancer::chooseHost(LoadBalancerContext*) {
-  const HostVector& hosts_to_use = hostSourceToHosts(hostSourceToUse());
-  if (hosts_to_use.empty()) {
-    return nullptr;
+HostConstSharedPtr RandomLoadBalancer::chooseHost(LoadBalancerContext* context) {
+  HostConstSharedPtr host;
+  size_t max_attempts = context ? context->hostSelectionRetryCount() + 1 : 1;
+  for (size_t i = 0; i < max_attempts; ++i) {
+    const HostVector& hosts_to_use = hostSourceToHosts(hostSourceToUse(context));
+    if (hosts_to_use.empty()) {
+      return nullptr;
+    }
+
+    host = hosts_to_use[random_.random() % hosts_to_use.size()];
+
+    // If host selection failed or the host is accepted by the filter, return.
+    // Otherwise, try again.
+    if (!host || !context || context->postHostSelectionFilter(*host)) {
+      return host;
+    } else {
+      continue;
+    }
```

Comment:
What if in LoadBalancerBase you write an implementation of `chooseHost(LoadBalancerContext*)` that handles the retry count and post host selection filtering in terms of a call to an abstract function, say `chooseHostOnce(LoadBalancerContext*)` (but feel free to come up with a better name) and then change the various load balancers to implement chooseHostOnce instead of chooseHost.

I think you'd be able to rename the existing chooseHost implementations to chooseHostOnce (with small changes) and keep the repeated retry/post filter logic in one location.